### PR TITLE
Introduce the Access List object.

### PR DESCRIPTION
**Pull request by mdwn**
The Access List object, which is the foundational object for access grants, has been introduced. Due to the size/complexity of this object, the implementation in api/types will come in a follow up.

[RFD-6E for context](https://github.com/gravitational/teleport.e/blob/master/rfd/0006e-access-lists.md)

### CODE REVIEW

**Comment by tigrato** [comment]
@mdwn do we have the reasoning behind this object anywhere?

It's quite difficult to review with no/little knowledge

**Comment by mdwn** [comment]
> @mdwn do we have the reasoning behind this object anywhere?
> 
> It's quite difficult to review with no/little knowledge

Ahh sorry, I meant to add in a link to the RFD in the PR description. It's been added.

**Comment by zmb3** [comment]
Thanks for breaking this change into a separate PR and for the great comments!

File: api/proto/teleport/legacy/types/types.proto [hunk 1]
```diff
@@ -6025,3 +6025,135 @@ message GCPMatcher {
...
+  // Description is a plaintext description of the access list.
+  string Description = 1 [(gogoproto.jsontag) = "description"];
+
+  // Owners is a list of owners of the access list.
+  repeated AccessListOwnerV1 Owners = 2 [(gogoproto.jsontag) = "owners"];
+
+  // Audit describes the frequency that this access list must be audited.
+  AccessListAuditV1 Audit = 3 [(gogoproto.jsontag) = "audit"];
+
+  // MembershipRequires describes the requirements for a user to be a member of the access list.
```
Comment (tigrato) [inline api/proto/teleport/legacy/types/types.proto:6061]: Should we consider adding a comment saying that for the access list to be effective, the user must be in members and satisfy `MembershipRequires`. The same for owners

File: api/proto/teleport/legacy/types/types.proto [hunk 2]
```diff
@@ -6025,3 +6025,135 @@ message GCPMatcher {
...
+
+  // Description is the plaintext description of the owner and why they are an owner.
+  string Description = 2 [(gogoproto.jsontag) = "description"];
+}
+
+// AccessListAuditV1 describes the audit configuration for an access list.
+message AccessListAuditV1 {
+  // Frequency is a string parseable by golang's time.ParseDuration that describes
+  // how often an access list must be audited.
+  string Frequency = 1 [(gogoproto.jsontag) = "frequency"];
```
[the remark this answers is not in the document]
Comment (tigrato) [reply api/proto/teleport/legacy/types/types.proto:6090]: +1 but for using well-known type [`Duration`](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/duration.proto)

**Review by tigrato — COMMENTED** [review]
I know what I'm going to ask for next is a lot of work but since you're the one who launched the RFD Gogo must go, can we try not to add more uses of Gogo?
Will we support syncing these objects or will they only exist on the auth server?

File: api/proto/teleport/legacy/types/types.proto [hunk 3]
```diff
@@ -6025,3 +6025,135 @@ message GCPMatcher {
...
+  ];
+}
+
+// AccessListSpecV1 is the specification for an access list.
+message AccessListSpecV1 {
+  // Description is a plaintext description of the access list.
+  string Description = 1 [(gogoproto.jsontag) = "description"];
+
+  // Owners is a list of owners of the access list.
+  repeated AccessListOwnerV1 Owners = 2 [(gogoproto.jsontag) = "owners"];
```
[the remark this answers is not in the document]
Comment (mdwn) [reply api/proto/teleport/legacy/types/types.proto:6056]: I suppose for maintaining parity with the rest of the codebase, or so I thought. On second look it looks like we don't typically do this. I'll go ahead and remove these!

**Comment by mdwn** [comment]
Hrm:
```
# github.com/gravitational/teleport/api/types
api/types/types.pb.go:38448:29: m.Frequency.MarshalToSizedBuffer undefined (type *durationpb.Duration has no field or method MarshalToSizedBuffer)
api/types/types.pb.go:47515:19: m.Frequency.Size undefined (type *durationpb.Duration has no field or method Size)
api/types/types.pb.go:105087:26: m.Frequency.Unmarshal undefined (type *durationpb.Duration has no field or method Unmarshal)
```
Gogo maybe doesn't like Duration?

**Comment by tigrato** [comment]
> Hrm:
> ```
> # github.com/gravitational/teleport/api/types
> api/types/types.pb.go:38448:29: m.Frequency.MarshalToSizedBuffer undefined (type *durationpb.Duration has no field or method MarshalToSizedBuffer)
> api/types/types.pb.go:47515:19: m.Frequency.Size undefined (type *durationpb.Duration has no field or method Size)
> api/types/types.pb.go:105087:26: m.Frequency.Unmarshal undefined (type *durationpb.Duration has no field or method Unmarshal)
> ```
> Gogo maybe doesn't like Duration?

You need to add a directive to the protoc compiler. 
I will check that today or tomorrow morning

**Comment by mdwn** [comment]
Okay, I've migrated this away from legacy into its own proto file. I had mistakenly thought that there would be an issue with the cache, but I was misremembering the reason that was a concern before.

**Comment by mdwn** [comment]
I've just added in a "common" package to create a header for this object -- I'm attempting to create this object in the "correct" way, so let me know if you have any objections and I can revert this commit.

**Review by zmb3 — APPROVED** [review]

**Review by tigrato — APPROVED** [review]

File: api/proto/teleport/common/v1/common.proto [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+syntax = "proto3";
+
+package teleport.common.v1;
+
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/gravitational/teleport/api/gen/proto/go/common/v1;common";
+
+// ResourceHeader is a shared resource header.
+message ResourceHeader {
```
Comment (codingllama) [inline api/proto/teleport/common/v1/common.proto:24]: nit: Use separate files? Importers of ResourceHeader don't necessarily need/want Metadata or Trait, and so on.

File: api/proto/teleport/common/v1/common.proto [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package teleport.common.v1;
```
Comment (codingllama) [inline api/proto/teleport/common/v1/common.proto:17]: I'm not sure a "common" package is a good idea, chances are it'll inflate over time.

**Review by codingllama — COMMENTED** [review]
Drive-by per Mike's request.

File: api/proto/teleport/common/v1/common.proto (hunk 5 above)
Comment (mdwn) [reply api/proto/teleport/common/v1/common.proto:17]: Do you have a suggestion? We do have a number of common needs -- `Traits`, `ResourceHeader`, and `Metadata` will be seen pretty commonly.

Comment (mdwn) [reply api/proto/teleport/common/v1/common.proto]: Talking on slack, used `header` and `trait` packages separately.

**Pull request merged**